### PR TITLE
Add MTY_Zoom and MTY_Cursor modules

### PR DESCRIPTION
**Pull request by Namaneo**
* `MTY_Zoom` module: This module provides out-of-the-box coordinates conversion and cursor positioning, based of on the `MTY_EVENT_MOTION` and `MTY_EVENT_SCALE` events' data. It supports multiple modes:

    * `MTY_INPUT_MODE_TOUCHSCREEN`: In this mode, the cursor position is directly transformed from the absolute screen coordinates to the zoomed viewport, and the image is never moved.
    * `MTY_INPUT_MODE_TRACKPAD`: In this mode, the cursor position is computed relatively from motion events, and the image is moved in case the cursor hits the borders of the zoomed viewport.
    * Relative: This special mode is made for when the input data are relative and the cursor position must not be converted.

* `MTY_Cursor` module: This helper module allows to create, setup and draw a cursor. This is basically a wrapper around `MTY_WindowDrawQuad`.

Requires:
* https://github.com/parsec-cloud/libmatoya/pull/6
* https://github.com/parsec-cloud/libmatoya/pull/8

### CODE REVIEW

File: src/cursor.c [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+#include <stdint.h>
+
+struct MTY_Cursor {
+	MTY_App *app;
+	bool enabled;
+	void *image;
+	uint32_t width;
+	uint32_t height;
+	uint16_t hotX;
+	uint16_t hotY;
```
Comment (RonaldH) [inline src/cursor.c]: Change to `hot_x` and `hot_y` and update comment regarding metal/dx12.

**Review by RonaldH — CHANGES_REQUESTED** [review]

File: src/cursor.c (hunk 1 above)
Comment (Namaneo) [reply src/cursor.c]: Added comment on `MTY_CursorDraw` in `matoya.h`

File: src/cursor.c [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+struct MTY_Cursor {
+	MTY_App *app;
+	bool enabled;
+	void *image;
+	uint32_t width;
+	uint32_t height;
+	uint16_t hot_x;
+	uint16_t hot_y;
+	int32_t x;
+	int32_t y;
```
Comment (RonaldH) [inline src/cursor.c]:
```suggestion
	float hot_x;
	float hot_y;
	float x;
	float y;
```

File: src/cursor.c [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+	desc.cropHeight = ctx->height;
+	desc.imageWidth = ctx->width;
+	desc.imageHeight = ctx->width;
+	desc.aspectRatio = (float) ctx->width / (float) ctx->height;
+	desc.blend = true;
+
+	desc.type = MTY_POSITION_FIXED;
+	desc.scale = ctx->scale;
+	desc.imageX = (int32_t) (ctx->x - (ctx->hot_x * desc.scale));
+	desc.imageY = (int32_t) (ctx->y - (ctx->hot_y * desc.scale));
```
Comment (RonaldH) [inline src/cursor.c]:
```suggestion
	desc.imageX = lrint(ctx->x - (ctx->hot_x * desc.scale));
	desc.imageY = lrint(ctx->y - (ctx->hot_y * desc.scale));
```

File: src/zoom.c [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+	return ctx && ctx->window_w && ctx->window_h && ctx->image_w && ctx->image_h;
+}
+
+static float mty_zoom_tranform_x(MTY_Zoom *ctx, float value)
+{
+	float offset_x = - ctx->image.x / ctx->scale_screen + ctx->image_min.x;
+	float zoom_w   = ctx->window_w / ctx->scale_screen;
+	float ratio_x  = (float) value / ctx->window_w;
+
+	return offset_x + zoom_w * ratio_x;
```
Comment (RonaldH) [inline src/zoom.c]:
```suggestion
	float offset_x = - ctx->image.x / ctx->scale_screen + ctx->image_min.x;
	float zoom_w   = (float) ctx->window_w / ctx->scale_screen;
	float ratio_x  = value / (float) ctx->window_w;

	return offset_x + zoom_w * ratio_x;
```

File: src/zoom.c [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+}
+
+static float mty_zoom_tranform_y(MTY_Zoom *ctx, float value)
+{
+	float offset_y = - ctx->image.y / ctx->scale_screen + ctx->image_min.y;
+	float zoom_h   = ctx->window_h / ctx->scale_screen;
+	float ratio_y  = (float) value / ctx->window_h;
```
Comment (RonaldH) [inline src/zoom.c]:
```suggestion
	float offset_y = -ctx->image.y / ctx->scale_screen + ctx->image_min.y;
	float zoom_h   = (float) ctx->window_h / ctx->scale_screen;
	float ratio_y  = value / (float) ctx->window_h;
```

**Review by RonaldH — CHANGES_REQUESTED** [review]

File: src/zoom.c [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+
+	return offset_y + zoom_h * ratio_y;
+}
+
+static void mty_zoom_restrict_image(MTY_Zoom *ctx)
+{
+	float image_scaled_w = ctx->image_w * ctx->scale_image;
+	float image_scaled_h = ctx->image_h * ctx->scale_image;
```
Comment (RonaldH) [inline src/zoom.c:91]:
```suggestion
	float image_scaled_w = (float) ctx->image_w * ctx->scale_image;
	float image_scaled_h = (float) ctx->image_h * ctx->scale_image;
```

File: src/zoom.c [hunk 7]
```diff
@@ -0,0 +1,387 @@
...
+		return;
+
+	ctx->window_w = windowWidth;
+	ctx->window_h = windowHeight;
+	ctx->image_w = imageWidth;
+	ctx->image_h = imageHeight;
+
+	ctx->scale_screen = 1;
+	float scale_w = (float) ctx->window_w / ctx->image_w;
+	float scale_h = (float) ctx->window_h / ctx->image_h;
```
Comment (RonaldH) [inline src/zoom.c]:
```suggestion
	float scale_w = (float) ctx->window_w / (float) ctx->image_w;
	float scale_h = (float) ctx->window_h / (float) ctx->image_h;
```

File: src/zoom.c [hunk 8]
```diff
@@ -0,0 +1,387 @@
...
+	ctx->scale_image = scale_w < scale_h ? scale_w : scale_h;
+
+	ctx->image.x = 0;
+	ctx->image.y = 0;
+	if (scale_w > scale_h) 
+		ctx->image.x = (ctx->window_w - ctx->image_w * ctx->scale_image) / 2.0f;
+	if (scale_w < scale_h) 
+		ctx->image.y = (ctx->window_h - ctx->image_h * ctx->scale_image) / 2.0f;
```
Comment (RonaldH) [inline src/zoom.c:130]:
```suggestion
	if (scale_w > scale_h) 
		ctx->image.x = ((float) ctx->window_w - (float) ctx->image_w * ctx->scale_image) / 2.0f;
	if (scale_w < scale_h) 
		ctx->image.y = ((float) ctx->window_h - (float) ctx->image_h * ctx->scale_image) / 2.0f;
```

File: src/zoom.c [hunk 9]
```diff
@@ -0,0 +1,387 @@
...
+	ctx->image_min.x = ctx->image.x;
+	ctx->image_min.y = ctx->image.y;
+	ctx->image_max.x = ctx->window_w - ctx->image.x;
+	ctx->image_max.y = ctx->window_h - ctx->image.y;
+
+	ctx->scale_image_min = ctx->scale_image * ctx->scale_screen_min;
+	ctx->scale_image_max = ctx->scale_image * ctx->scale_screen_max;
+
+	ctx->cursor.x = ctx->window_w / 2.0f;
+	ctx->cursor.y = ctx->window_h / 2.0f;
```
Comment (RonaldH) [inline src/zoom.c:141]:
```suggestion
	ctx->cursor.x = (float) ctx->window_w / 2.0f;
	ctx->cursor.y = (float) ctx->window_h / 2.0f;
```

File: src/zoom.c [hunk 10]
```diff
@@ -0,0 +1,387 @@
...
+
+	ctx->margin = (ctx->window_w < ctx->window_h ? ctx->window_w : ctx->window_h) * 0.2f;
```
Comment (RonaldH) [inline src/zoom.c:143]:
```suggestion
	ctx->margin = (float) (ctx->window_w < ctx->window_h ? ctx->window_w : ctx->window_h) * 0.2f;
```

File: src/zoom.c [hunk 11]
```diff
@@ -0,0 +1,387 @@
...
+	assert(ctx, false);
+
+	return ctx->mode == MTY_INPUT_MODE_TRACKPAD;
+}
+
+void MTY_ZoomEnableTrackpad(MTY_Zoom *ctx, bool enable)
+{
+	assert(ctx);
+
+	ctx->mode = enable ? MTY_INPUT_MODE_TRACKPAD : MTY_INPUT_MODE_TOUCHSCREEN;
```
Comment (RonaldH) [inline src/zoom.c:360]: Should this be future proofed, and remove the flags first? (something like
```c
ctx->mode = ctx->mode & ^MTY_INPUT_MODE_TRACKPAD & ^MTY_INPUT_MODE_TOUCHSCREEN;
ctx->mode |= enable ? MTY_INPUT_MODE_TRACKPAD : MTY_INPUT_MODE_TOUCHSCREEN;
```

File: src/zoom.c [hunk 12]
```diff
@@ -0,0 +1,387 @@
...
+}
+
+int32_t MTY_ZoomGetCursorY(MTY_Zoom *ctx)
+{
+	assert(ctx, 0);
+
+	float top    = mty_zoom_tranform_y(ctx, 0);
+	float bottom = mty_zoom_tranform_y(ctx, (float) ctx->window_h);
+
+	return (int32_t) (ctx->window_h * (ctx->cursor.y - top) / (bottom - top));
```
Comment (RonaldH) [inline src/zoom.c]: lrint and (float)

File: src/zoom.c [hunk 13]
```diff
@@ -0,0 +1,387 @@
...
+bool MTY_ZoomShouldShowCursor(MTY_Zoom *ctx)
+{
+	assert(ctx, false);
+
+	return ctx->mode == MTY_INPUT_MODE_TRACKPAD && !ctx->relative;
+}
+
+void MTY_ZoomSetLimits(MTY_Zoom *ctx, float min, float max)
+{
+	assert(ctx);
```
Comment (RonaldH) [inline src/zoom.c]: Remove all asserts.

**Review by RonaldH — CHANGES_REQUESTED** [review]

File: src/zoom.c (hunk 11 above)
Comment (Namaneo) [reply src/zoom.c:360]: The `MTY_InputMode` values are not meant to be used as flags (I guess)

**Review by RonaldH — APPROVED** [review]

File: src/cursor.c [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+	MTY_Cursor *ctx = MTY_Alloc(1, sizeof(MTY_Cursor));
+
+	ctx->app = app;
+	ctx->window = window;
+	ctx->enabled = true;
+
+	ctx->commands = MTY_Alloc(CMD_LEN, sizeof(MTY_Cmd));
+	ctx->vertices = MTY_Alloc(VTX_LEN, sizeof(MTY_Vtx));
+	ctx->indices  = MTY_Alloc(IDX_LEN, sizeof(uint16_t));
+	ctx->cmd_list = MTY_Alloc(1, sizeof(MTY_CmdList));
```
Comment (RonaldH) [inline src/cursor.c]: Just put this directly in the command, and free the command if available.

**Review by RonaldH — CHANGES_REQUESTED** [review]

**Review by Kodikuu — APPROVED** [review]
I don't know how I feel about `VALIDATE_CTX`, but otherwise this looks fine

**Review by Kodikuu — APPROVED** [review]

**Comment by Namaneo** [comment]
Theorically useless for now, closing.

**Pull request closed without merging**